### DISCUSSION TITLE
Adding multiple domains

### DISCUSSION THREAD

**Discussion by PrieserMax** [simple-login/app#1987] [Self-hosting question]
Heyo,
I would like to add more domains.
How is that possible?
I can imagine, adding them to the .env file, expanding the SSL-certficate. But what is with Postfix?
I'm not that sure (like here: #1912 )
Thanks for the great work so far, this service is amazing and being able to use it with Bitwarden (Vaultwarden) is perfect!

Greetings
Max

**Comment by BJReplay** [comment]
I've just figured this out from another discussion / issue:

Caveats:
- Running with a wildcard certificate so didn't need to think about that - but if you've got automatic certificates then that will presumably be taken care of, otherwise you'll need to replace your certificate.
- Running with a modified version of simplelogin/postfix docker container (modified to deal with the fact that spamhaus blocks queries from recursive DNS servers, so I'm using DQN, and I modified docker to pull in the DQN key from config and update main.cf with this.  You probably don't need to worry about this - just including for full disclosure.
- modified simplelogin/postfix docker container to accept and use RELAY_PORT.  You probably don't need to worry about this - just including for full disclosure.

What I found worked for me:
- I had previously set up a custom domain in the UI, so I copied the three DKIM CNAME entries, the SPF TXT record and the DMARC TXT record for the custom domain to my new intended public domain DNS entries, as well as an MX record pointing back to my main simplelogin domain (so all in all, six DNS records for my new domain.
- I was adding two new domains, so did this (added six DNS entries) for my second domain.
- I added both new domains to `OTHER_ALIAS_DOMAINS=["email.domain1.tld", "email.domain2.tld"]` in my .env file
- I restarted all docker containers (`docker compose down && docker compose up -d && docker compose exec sl-db psql -U postgres -d simplelogin`)
- You might get away with restarting less containers, but restarting the stack makes sense, and I know it worked.
- `select * from public_domain;` confirmed both new domains were in and fully set up.

If you haven't previously set up custom domains, the DNS entries you'll need set up are:

Assuming setting up `email.domain.tld`

- `CNAME dkim._domainkey.email dkim._domainkey.<your main simplelogin domain here>`
- `CNAME dkim02._domainkey.email dkim02._domainkey.<your main simplelogin domain here>`
- `CNAME dkim03._domainkey.email dkim03._domainkey.<your main simplelogin domain here>`
- `TXT email v=spf1 include:<your main simplelogin domain here>~all`
- `TXT _dmarc.email v=DMARC1; p=quarantine; pct=100; adkim=s; aspf=s` (no customisation required)
- `MX email <your main simplelogin domain here> 10` (default TTL is fine)

If just setting up `domain.tld`

- `CNAME dkim._domainkey dkim._domainkey.<your main simplelogin domain here>`
- `CNAME dkim02._domainkey dkim02._domainkey.<your main simplelogin domain here>`
- `CNAME dkim03._domainkey dkim03._domainkey.<your main simplelogin domain here>`
- `TXT @ v=spf1 include:<your main simplelogin domain here>~all`
- `TXT _dmarc v=DMARC1; p=quarantine; pct=100; adkim=s; aspf=s` (no customisation required)
- `MX @ <your main simplelogin domain here> 10` (default TTL is fine)

**Comment by brettgus** [comment]
@BJReplay  Didn't you have to modify postfix at all to receive email for the other domain?  Otherwise it would just see emails to the additional domain as relaying?

**Comment by BJReplay** [comment]
@brettgus quite possibly - it's been a while 🤣.

From what I can see, there's nothing in my docker compose for this, but in my `.env.sl`, I have

`OTHER_ALIAS_DOMAINS=["other.domain.first", "other.domain.second"]` 

So, I'm guessing that that's the step I left out of the instructions above

Docker compose under my postfix container just has 
`ALIASES_DEFAULT_DOMAIN: main.domain.here`

P.S. since setting it up, aside from having to rebuild the docker images myself (since I run on oracle cloud free on arm architecture), this stuff just works.

**Comment by brettgus** [comment]
Ahh, one difference is that I'm using the instructions from the README to install postfix locally on my machine, it sounds like you're using the docker container postfix.

I figured out what the problem was though, in case anyone else runs into this issue.

The configuration in the README recommended for /etc/postfix/pgsql-relay-domains.cf is 
```
# postgres config
hosts = localhost
user = myuser
password = mypassword
dbname = simplelogin

query = SELECT domain FROM custom_domain WHERE domain='%s' AND verified=true
    UNION SELECT '%s' WHERE '%s' = 'mydomain.com' LIMIT 1;
```

And for /etc/postfix/pgsql-transport-maps.cf is 
```
# postgres config
hosts = localhost
user = myuser
password = mypassword
dbname = simplelogin

# forward to smtp:127.0.0.1:20381 for custom domain AND email domain
query = SELECT 'smtp:127.0.0.1:20381' FROM custom_domain WHERE domain = '%s' AND verified=true
    UNION SELECT 'smtp:127.0.0.1:20381' WHERE '%s' = 'mydomain.com' LIMIT 1;
```

However, neither of those accomodate for public_domains, only custom_domains.

Tweaking them respectively to add public_domains fixed the issue:
```
query = SELECT domain FROM custom_domain WHERE domain='%s' AND verified=true
    UNION SELECT domain FROM public_domain WHERE domain='%s'
    UNION SELECT '%s' WHERE '%s' = 'mydomain.com' LIMIT 1;
```

```
query = SELECT 'smtp:127.0.0.1:20381' FROM custom_domain WHERE domain='%s' AND verified=true
    UNION SELECT 'smtp:127.0.0.1:20381' FROM public_domain WHERE domain='%s'
    UNION SELECT 'smtp:127.0.0.1:20381' WHERE '%s' = 'mydomain.com' LIMIT 1;
```
